### ISSUE TITLE
kvserver: snapshot ingestion rates are not bounded

### ISSUE THREAD

**Issue by irfansharif** [cockroachdb/cockroach#74694]
**Is your feature request related to a problem? Please describe.**

We have cluster settings that control how quickly rebalance/recovery snapshots are generated: `kv.snapshot_{rebalance,recovery}.max_rate`. What we're missing is any sort of control for how quickly snapshots can be ingested. For even medium sized clusters (say, 50 nodes), our default recovery rates of [32 MB/s](https://github.com/cockroachdb/cockroach/pull/71814), various cluster events (decommissioning/outages/zone config changes) could end up hammering specific nodes with snapshots much quicker than it's equipped to handle. We saw evidence of this in a [recent escalation](https://github.com/cockroachlabs/support/issues/1374); this can manifest as burst of writes to the node's underlying LSM possibly inverting it. The inversion can then lead to high read-amps, disruptive to the node's handling of foreground traffic.

**Describe the solution you'd like**

It seems bad that a node is at mercy of sender rates, which if sufficiently high, can end up toppling it quite easily. Some form of crude receiver side throttling could perhaps help with flow control. Few options:
- through a symmetric `kv.snapshot_{rebalance,recovery}_receiver.max_rate`;
- (from @nvanbenschoten) snapshot ingestion is not subject to Pebble’s PreIngestDelay like (above and below Raft) AddSSTable is. I don’t think that was intentional and it may serve as a kind of snapshot receiver-side throttling.

+cc @cockroachdb/kv-notifications.

**Comment by nvanbenschoten** [comment]
> through a symmetric kv.snapshot_{rebalance,recovery}_receiver.max_rate;

I don't understand what this gives us. We already limit the receiver-side concurrency to 1, so the effective rate limit would be the minimum of the sender and receiver limits. Presumably, customers that increase the sender-side rate limit will also increase the receiver-side rate limit accordingly.

**Comment by irfansharif** [comment]
With a receiver-side concurrency of 1, what would you say is our snapshot ingestion rate in terms of `snapshot_rebalance.max_rate`? When thinking of throttling, I meant any kind that would reduce the likelihood/prevent inverted LSMs after a node receives a lot of snapshot data.

> snapshot ingestion is not subject to Pebble’s PreIngestDelay like (above and below Raft) AddSSTable is. I don’t think that was intentional and it may serve as a kind of snapshot receiver-side throttling

I need to remind myself how snapshots work, but they also add SSTables, right? You're saying that particular code path skips Pebble's PreIngestDelay?

**Comment by nvanbenschoten** [comment]
> I need to remind myself how snapshots work, but they also add SSTables, right? You're saying that particular code path skips Pebble's PreIngestDelay?

Yes, snapshots add SSTables and use the `IngestExternalFiles` API to add them to the LSM. However, no one in the snapshot receiver code path calls `PreIngestDelay`, which sits off to the side and is not part of `IngestExternalFiles` API directly.

**Comment by tbg** [comment]
Closing in favor of https://github.com/cockroachdb/cockroach/issues/80607.
